feat(common): add loaders for Cloudinary and ImageKit providers

khempenius opened the pull request:
## PR Type
What kind of change does this PR introduce?
- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

khempenius committed:
feat(common): add loaders for cloudinary & imagekit

This commit adds loaders for cloudinary and imagekit.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudinary_loader.ts:
```diff
@@ -0,0 +1,75 @@
+  };
+}
+
+function assertValidPath(path: unknown) {
```
The code from here down to the end of this file is duplicated between loaders (similar code already [exists in the repo already](https://github.com/angular/angular/blob/7173711d16668003a41f095c615e39ae125f760a/packages/common/src/directives/ng_optimized_image/image_loaders/imgix_loader.ts#L39-L67)).

We should extract those utilities into something like loader_utils.ts or similar.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/imagekit_loader.ts:
```diff
@@ -0,0 +1,74 @@
+  };
+}
+
+function assertValidPath(path: unknown) {
```
The code from here down to the end of this file is duplicated between loaders (similar code already [exists in the repo already](https://github.com/angular/angular/blob/7173711d16668003a41f095c615e39ae125f760a/packages/common/src/directives/ng_optimized_image/image_loaders/imgix_loader.ts#L39-L67)).

We should extract those utilities into something like loader_utils.ts or similar.

pkozlowski-opensource reviewed (changes requested):
The change LGTM overall but we start to have code duplication between various loaders and I would love to see the common logic extracted before this PR gets merged.

kara commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudinary_loader.ts:
```diff
@@ -0,0 +1,54 @@
+    useValue: (config: ImageLoaderConfig) => {
+      // Example of a Cloudinary image URL:
+      // https://res.cloudinary.com/mysite/image/upload/c_scale,f_auto,q_auto,w_600/marketing/tile-topics-m.png
+      let params = `f_auto,q_auto`;
```
Can we add a brief comment for what these options mean and why we're adding them?

kara commented on packages/common/src/directives/ng_optimized_image/image_loaders/imagekit_loader.ts:
```diff
@@ -0,0 +1,53 @@
+    useValue: (config: ImageLoaderConfig) => {
+      // Example of an ImageKit image URL:
+      // https://ik.imagekit.io/demo/tr:w-300,h-300/medium_cafe_B1iTdD0C.jpg
+      let params = `tr:q-auto`;
```
Comment for what this param does?

kara reviewed (approved):
LGTM, minor nits below

AndrewKushnir committed:
refactor(common): allow Cloudinary and ImageKit loaders to specify the ensurePreconnect option

This commit modifies the provideCloudinaryLoader and provideImageKitLoader functions to support an additional ensurePreconnect option, similar to the Imgix loader functions.

AndrewKushnir committed:
refactor(common): move loader util functions to a common location

This commit refactors the code to avoid an extra utils file and instead, all extra helpers are moved to a common location.

pkozlowski-opensource reviewed (approved):
LGTM
